lib/File/Which.pm

```diff
@@ -99,33 +99,86 @@ me.
 our @EXPORT    = 'which';
 our @EXPORT_OK = 'where';
 
-use constant IS_VMS => ($^O eq 'VMS');
-use constant IS_MAC => ($^O eq 'MacOS');
-use constant IS_WIN => ($^O eq 'MSWin32' or $^O eq 'dos' or $^O eq 'os2');
-use constant IS_DOS => IS_WIN();
-use constant IS_CYG => ($^O eq 'cygwin' || $^O eq 'msys');
-
-our $IMPLICIT_CURRENT_DIR = IS_WIN || IS_VMS || IS_MAC;
-
-# For Win32 systems, stores the extensions used for
-# executable files
-# For others, the empty string is used
-# because 'perl' . '' eq 'perl' => easier
-my @PATHEXT = ('');
-if ( IS_WIN ) {
-  # WinNT. PATHEXT might be set on Cygwin, but not used.
-  if ( $ENV{PATHEXT} ) {
-    push @PATHEXT, split /;/, $ENV{PATHEXT};
+sub _get_osname { @_ == 1 && ref $_[0] ? $_[0]->{osname} : $^O }
+
+sub _is_vms { my $osname = &_get_osname; ($osname eq 'VMS'); }
+sub _is_mac { my $osname = &_get_osname; ($osname eq 'MacOS'); }
+sub _is_win { my $osname = &_get_osname; ($osname eq 'MSWin32' or $osname eq 'dos' or $osname eq 'os2'); }
+sub _is_dos { _is_win(@_); }
+sub _is_cyg { my $osname = &_get_osname; ($osname eq 'cygwin' || $osname eq 'msys'); }
+
+sub _default_implicit_current_dir {
+  my $self = shift;
+  $self->_is_win || $self->_is_vms || $self->_is_mac;
+}
+our $IMPLICIT_CURRENT_DIR = do {
+  File::Which->new->_default_implicit_current_dir;
+};
+
+sub new {
+  my ($class, %opts) = @_;
+
+  my $osname = exists $opts{os} ? $opts{os} : $^O;
```

Comment:
I'm not sure `$^O` is quite right for this, it has a lot of detail for unix like operating systems and not enough for windows.  I think we want to know if it is unix, or win/vms/cyg/os9, and then we want something like win+powershell win+bash.

Reply:
Maybe we can have opts for `os` and `shell`.  I don't think `$^O` is right for `os` as Windows9x and WindowsNT are both `MSWin32` unfortunately.  I'm not too enthused about inventing our own categories for os, but we may have to.